# LeetCode Scramble

За задачку садился в течении нескольких вечеров по 2ч+-. Кроме контр примеров и некоторых их свойст, ничего интересного не установил. Принял решение сделать не оптимальный алгоритм решающий через поиск по всем возможным вариантам.

1. Строю древо всех возможных по условию задачи комбинаций.
2. Из древа получаю множество уникальных комбинаций.
3. Ищу по множеству требуемую комбинацию.

Думаю решение будет очень не оптимальное зато короткое. Далее буду оптимизировать и\или искать другое решение.

## Решение 1: Поиск $S_2$ среди всех доступных комбинаций. 

In [1]:
import itertools as it
import math as ma

def mtl(s):
    ts = set([s,])
    if(len(s) == 1):
        return ts
    else:
        for i in range(len(s) - 1):
            Sa = mtl(s[0:i+1])
            Sb = mtl(s[i+1:len(s)])
            ts.update({*(''.join(t) for t in it.product(Sa, Sb)), *(''.join(t) for t in it.product(Sb, Sa))})
        return ts

In [2]:
t = mtl('abcd')
print(len(t), t)
print('cadb' in t)
print('bdac' in t)

22 {'abcd', 'acdb', 'acbd', 'cabd', 'cdab', 'cbad', 'dacb', 'dcab', 'dabc', 'dbac', 'adbc', 'bcda', 'abdc', 'badc', 'cbda', 'bdca', 'bacd', 'cdba', 'dcba', 'bcad', 'dbca', 'adcb'}
False
False


22 комбинации из 24 возможных, отсутсвует $cadb$ и зеркальное $bdac$. На данном этапе понятно, что можно представить комбинациии букв как комбинации их индексов в исходной строке, тогда недоступные комбинации это комбинации вида - численный интервал $(a,d)$ из $N+<|S|$ переставленный в вид $c,a,d,b$. И если в комбинации присуствует любой такой интервал с маской  $*c*a*d*b*$ или зеркальный. Для букв встречающихся больше 1 раза, рассмотреть каждую возможную комбинацию индексов. Далее задачу буду решать в этом ключе.

Первое решение не прошло по time limit. На вот этих строчках.

In [3]:
#s1 = "abcdefghijklmn"
#s2 = "efghijklmncadb"
#len(s1)
#%time t = mtl(s1)
#s2 in t

У меня на компьютере занимает 3 минуты, сервер давал секунд 20. Видимо такая ветка решений вообще тупиковая по быстродействию.

## Решение 2: Поиск  𝑆2  среди всех недоступных комбинаций.

Строю алгоритм формирующий все тупиковые комбинации индексов из строки длинной $n$ $(3<n<31)$, пока оставляю все буквы уникальными. Потом буду решать вопрос с расстановкой индексов в строках $S_2$ когда букве может соответствовать больше 1 индекса прим. $cadbad$, логика такая, если хотя бы одна расстановка индексов достижима то вся комбинация достижима.

In [4]:
#2031 - 3142
s1 = 'abcde'
s2 = 'cadbe'

t = mtl(s1)
print(f' \nВсего возможных комбинаций - {ma.factorial(len(s1))} \nКол-во достижимых комбинаций - {len(t)}  \n\nДостижимые комбинации - {t}')

 
Всего возможных комбинаций - 120 
Кол-во достижимых комбинаций - 90  

Достижимые комбинации - {'dbcae', 'cedab', 'abedc', 'abdec', 'badec', 'dbcea', 'cbdea', 'bdeca', 'ecdba', 'bcdae', 'adcbe', 'bdcae', 'ecbda', 'cdbea', 'ebdca', 'cdeab', 'cbeda', 'abcde', 'badce', 'cabde', 'ebacd', 'aedbc', 'baced', 'bdcea', 'cbdae', 'dceab', 'acdbe', 'adbce', 'baedc', 'decba', 'baecd', 'dbace', 'dceba', 'ebcda', 'cbaed', 'debca', 'acbed', 'dacbe', 'cdeba', 'bcade', 'dcabe', 'eabcd', 'ecbad', 'cabed', 'abdce', 'edbac', 'adceb', 'dabce', 'adecb', 'abced', 'debac', 'ebcad', 'bedca', 'becda', 'bcaed', 'ecabd', 'dcbae', 'adebc', 'bcdea', 'decab', 'edacb', 'acedb', 'edbca', 'eabdc', 'eacdb', 'dcbea', 'ebadc', 'aecdb', 'ecdab', 'cbade', 'aebcd', 'eadcb', 'cdbae', 'cdabe', 'deacb', 'eacbd', 'edabc', 'acdeb', 'acbde', 'eadbc', 'aebdc', 'deabc', 'bacde', 'edcab', 'aedcb', 'abecd', 'cedba', 'aecbd', 'edcba', 'bceda'}


In [5]:
t = mtl(s1)
print(f' \nВсего возможных комбинаций - {ma.factorial(len(s1))} \nКол-во достижимых комбинаций - {len(t)}  \n\nДостижимые комбинации - {t}')

 
Всего возможных комбинаций - 120 
Кол-во достижимых комбинаций - 90  

Достижимые комбинации - {'dbcae', 'cedab', 'abedc', 'abdec', 'badec', 'dbcea', 'cbdea', 'bdeca', 'ecdba', 'bcdae', 'adcbe', 'bdcae', 'ecbda', 'cdbea', 'ebdca', 'cdeab', 'cbeda', 'abcde', 'badce', 'cabde', 'ebacd', 'aedbc', 'baced', 'bdcea', 'cbdae', 'dceab', 'acdbe', 'adbce', 'baedc', 'decba', 'baecd', 'dbace', 'dceba', 'ebcda', 'cbaed', 'debca', 'acbed', 'dacbe', 'cdeba', 'bcade', 'dcabe', 'eabcd', 'ecbad', 'cabed', 'abdce', 'edbac', 'adceb', 'dabce', 'adecb', 'abced', 'debac', 'ebcad', 'bedca', 'becda', 'bcaed', 'ecabd', 'dcbae', 'adebc', 'bcdea', 'decab', 'edacb', 'acedb', 'edbca', 'eabdc', 'eacdb', 'dcbea', 'ebadc', 'aecdb', 'ecdab', 'cbade', 'aebcd', 'eadcb', 'cdbae', 'cdabe', 'deacb', 'eacbd', 'edabc', 'acdeb', 'acbde', 'eadbc', 'aebdc', 'deabc', 'bacde', 'edcab', 'aedcb', 'abecd', 'cedba', 'aecbd', 'edcba', 'bceda'}


In [6]:
s2 in t

False

На строках длинной 5 алгоритм должен дать 30 недостижимых комбинаций. 

In [7]:
s1 = 'abcdefgh'
s2 = 'cadbefgh'

t = mtl(s1)
fn = ma.factorial(len(s1))
an = len(t)

print(f' \nВсего возможных комбинаций - {fn} \nКол-во достижимых комбинаций - {an}  \n\nДостижимые комбинации - слшиком длинный вывод ')
print(f'Кол-во недостижимых комбинаций - {fn - an}')

 
Всего возможных комбинаций - 40320 
Кол-во достижимых комбинаций - 8558  

Достижимые комбинации - слшиком длинный вывод 
Кол-во недостижимых комбинаций - 31762


Кол-во недостижимых комбинаций, очень быстро растет с ростом n при $n=5$ $u=30$ при $n=8$ $u=32000$, но мне нужны только те из них что длинной 4 их должно быть не много. Т.к. любая недостижимая комбинация большая 4, должна содержать в себе недостижимую комбинацию длинной 4.

Кол-во недостижимых комбинаций длинной $4$ это кол-во интервалов длинной длинной $4$ от $0$ до $n$, где n длинна строки т.е $n-4$ и еще столько же зеркальных комбинаций $(n-4)*2$. Это нужно проверить на небольшом множестве.

In [8]:
s1 = 'abcde'
s2 = 'cadbe'

t = mtl(s1)
fn = ma.factorial(len(s1))
an = len(t)

print(f' \nВсего возможных комбинаций - {fn} \nКол-во достижимых комбинаций - {an}  \n\nДостижимые комбинации - слшиком длинный вывод ')
print(f'Кол-во недостижимых комбинаций - {fn - an}')

all_comb = {''.join(c) for c in it.permutations([*s1])}
unavailable_comb = all_comb.symmetric_difference(t)

 
Всего возможных комбинаций - 120 
Кол-во достижимых комбинаций - 90  

Достижимые комбинации - слшиком длинный вывод 
Кол-во недостижимых комбинаций - 30


In [9]:
#cadb - bdac
unavailable_comb

{'acebd',
 'adbec',
 'bcead',
 'bdace',
 'bdaec',
 'bdeac',
 'beacd',
 'beadc',
 'becad',
 'bedac',
 'cadbe',
 'cadeb',
 'caebd',
 'caedb',
 'cbead',
 'cdaeb',
 'ceabd',
 'ceadb',
 'cebad',
 'cebda',
 'dabec',
 'daceb',
 'daebc',
 'daecb',
 'dbaec',
 'dbeac',
 'dbeca',
 'dcaeb',
 'ebdac',
 'ecadb'}

Для строки длинной $5$ получается $2$ интервала $a,b,c,d$ и $b,c,d,e$ длинной $4$ каждый интервал имеет зеркальный интервал. Всего $4$ интервала которые можно спутать переставив символы как 
$0,1,2,3$ -> $2,0,3,1$ из каждого недостижимого интервала можно получить комбинации добавляя все комбинации остальных свободных букв(индексов) между букв и с концов по схеме $*2*3*0*1*$. Получается $20$ комбинаций. Не понятно откуда еще $10$. Надо проверять.

In [10]:
len(unavailable_comb)

30

Лишние 10 строк это новый вид недоступных комбинаций вида $1403(3041)$ индексы обладают теми же свойствами, что $2031$, но не являются числами одного интервала, значит условие про итервал неверное. Принцип формирующий паттерны описанный ранее неполный или неверный. 

In [11]:
s1 = 'abcdef'
s2 = 'cadbef'

t = mtl(s1)
fn = ma.factorial(len(s1))
an = len(t)

print(f' \nВсего возможных комбинаций - {fn} \nКол-во достижимых комбинаций - {an}  \n\nДостижимые комбинации - слшиком длинный вывод ')
print(f'Кол-во недостижимых комбинаций - {fn - an}')

all_comb = {''.join(c) for c in it.permutations([*s1])}
unavailable_comb = all_comb.symmetric_difference(t)

 
Всего возможных комбинаций - 720 
Кол-во достижимых комбинаций - 394  

Достижимые комбинации - слшиком длинный вывод 
Кол-во недостижимых комбинаций - 326


При $n=6$ кол-во недостижимых комбинаций из $30$ становится $326$. Считаю по старому паттерну, умножаю кол-во недоступных комб. длинной $4$ на $n+1$ мест куда можно расставить остальные свободные буквы и умножаю на кол-во перестановок свободных букв.

In [12]:
n = 6
u_4_base = 4*ma.factorial(n-3)
u_patterns_n = u_4_base*2*ma.factorial(n-4)*5
print(u_patterns_n)

480


Результат совершенно далекий от правды, видимо не учивтываю какие то пересечения в формуле. На данный момент подсчет не имеет значения, потому что генерация всех недоступных комбинаций выглядит слишком времязатратной.

Предполагаю, что общее у всех недоступных четверок при длинне строки $(3<n<31)$ это соответсвие паттерну $i_2>i_0>i_3>i_1$,где $i_n$ это число стоящее на позиции с индексом n. Прим. $S_1$ = $abcde$ по индексам $S_1^i = 01234$, пример недоступной строки согласно паттерну, сначало второй по величине индекс $i_0 = 3$, затем самый маленький $i_1= 0, i_2= 4, i_3=1$, получаем $3041, 2031$ так же подходит. Решение буду строить на поиске такого патерна в строке $S_2$. Принял решение на основании того, что недоступных комбинаций после $n=6$, становится больше чем доступных и поиск $S_2$ среди недоступных будет еше неоптимальнее. Так генерации всех недоступны комбинаций затратит много времени. Т.к это противоречит тому, что полагалось в начале решения 2. Это решение будет под номером 3.

## Решение 3: Поиск паттерна $i_2>i_0>i_3>i_1$ в строке $S_2$

По условию максимальная длинна строки 30. Такой поиск это проверка для каждого символа наличия условий из паттерна. Даже при грубом подсчете $O(n) = 30*30*op\_count*letters\_variance\_count$ выглядит неплохо относительно предыдущих решений.

In [86]:
def is_unavailable_debug(s):
    n=len(s)
    for i in range(n):
        if s[i] - 1 > 0: #Есть раcстояние для i1, i3
            for j in range(i+1,n):
                if s[i] - s[j] > 1: # Есть раcстояние для  i3
                    for k in range(j+1,n):
                        if s[k] > s[i]: # i2 Больше i0.
                            for h in range(k+1,n):
                                if s[h] > s[j] and s[h] < s[i]: # i3 > i1
                                    print(s[i],s[j],s[k], s[h])
                                    return True # Одного такого паттерна достаточно, что вся комбинация оказалась недостижима  
    s.reverse()
    for i in range(n):
        if s[i] - 1 > 0: #Есть раcстояние для i1, i0
            for j in range(i+1,n):
                if s[i] - s[j] > 1: # Есть раcстояние для  i3
                    for k in range(j+1,n):
                        if s[k] > s[i]: # i2 Больше i0.
                            for h in range(k+1,n):
                                if s[h] > s[j] and s[h] < s[i]: # i3 > i1 и i3 < i0
                                    print(s[i],s[j],s[k], s[h])
                                    return True # Одного такого паттерна достаточно, что вся комбинация оказалась недостижима 
    return False

In [87]:
def is_unavailable(s):
    n=len(s)
    for i in range(n):
        if s[i] - 1 > 0: #Есть раcстояние для i1, i0
            for j in range(i+1,n):
                if s[i] - s[j] > 1: # Есть раcстояние для  i3
                    for k in range(j+1,n):
                        if s[k] > s[i]: # i2 Больше i0.
                            for h in range(k+1,n):
                                if s[h] > s[j] and s[h] < s[i]: # i3 > i1
                                    #print(s[i],s[j],s[k], s[h])
                                    return True # Одного такого паттерна достаточно, что вся комбинация оказалась недостижима  
    s.reverse()
    for i in range(n):
        if s[i] - 1 > 0: #Есть раcстояние для i1, i0
            for j in range(i+1,n):
                if s[i] - s[j] > 1: # Есть раcстояние для  i3
                    for k in range(j+1,n):
                        if s[k] > s[i]: # i2 Больше i0.
                            for h in range(k+1,n):
                                if s[h] > s[j] and s[h] < s[i]: # i3 > i1
                                    #print(s[i],s[j],s[k], s[h])
                                    return True # Одного такого паттерна достаточно, что вся комбинация оказалась недостижима 
    return False

In [88]:
t = [0,1,2,3,4,5]
s = [2,0,3,1,4,5]

print(is_unavailable(s))
print(is_unavailable(t))

True
False


Теперь нужно написать алгорит конвертации из букв в индексы и подсчета вариантов расстановки индкесов для одинаковых букв. Затем проверить алгоритмы на имеющихся множествах.

In [89]:
def to_s1_index(s1,s2):
    index_dict = dict(zip([*s1],range(len(s1))))
    s = list(map(lambda x: index_dict[x], [*s2]))
    return s

In [90]:
s1 = 'abcd'
s2 = 'cadb'

to_s1_index(s1,s2)

[2, 0, 3, 1]

Далее тестирую алгоритм решения, беру множество всех доступных комбинаций и для них он даст False и множество всех недоступных, для них True.

In [107]:
s1 = 'abcdef'

all_comb = {''.join(c) for c in it.permutations([*s1])}
available_comb = mtl(s1)
unavailable_comb = all_comb.symmetric_difference(available_comb)

In [108]:
a_res = [(to_s1_index(s1, comb), is_unavailable(to_s1_index(s1, comb))) for comb in available_comb]
ua_res = [(to_s1_index(s1, comb), is_unavailable(to_s1_index(s1, comb))) for comb in unavailable_comb]

Среди доступных обноружились недоступные, это не верно, буду отлаживать.

In [109]:
tuple(filter(lambda x: x[1] == True , a_res))

()

In [110]:
is_unavailable_debug([2, 0, 1, 3, 4])

False

Все недоступные, недоступны это верно.

In [111]:
all(ua_res)

True

Остался вопрос со строками вида $aaaabcd$